cli/src/commands/simplify.rs

```diff
@@ -0,0 +1,133 @@
+    /// (can be repeated)
+    #[arg(long, short)]
+    sources: Vec<RevisionArg>,
+    /// Simplify specified revision(s) (can be repeated)
+    #[arg(long, short)]
+    revisions: Vec<RevisionArg>,
+}
+
+pub(crate) fn cmd_simplify(
+    ui: &mut Ui,
+    command: &CommandHelper,
+    args: &SimplifyArgs,
+) -> Result<(), CommandError> {
+    if args.sources.is_empty() && args.revisions.is_empty() {
+        return Err(user_error("no revisions specified"));
+    }
+
+    let mut workspace_command = command.workspace_helper(ui)?;
+    let revs = RevsetExpression::descendants(
+        workspace_command
+            .parse_union_revsets(&args.sources)?
+            .expression(),
+    )
+    .union(
+        workspace_command
+            .parse_union_revsets(&args.revisions)?
+            .expression(),
+    )
+    .minus(
+        &revset_util::parse_immutable_heads_expression(&workspace_command.revset_parse_context())?
+            .ancestors(),
+    );
+    let commits: Vec<_> = workspace_command
+        .attach_revset_evaluator(revs)
+        .evaluate_to_commits()?
+        .try_collect()?;
+    let orig_commits = commits.len();
+
+    let mut tx = workspace_command.start_transaction();
+    let mut stats = SimplifyStats::default();
+    for commit in commits {
+        stats.add(simplify_commit(command.settings(), &mut tx, &commit)?);
+    }
+
+    if let Some(mut formatter) = ui.status_formatter() {
+        if !stats.is_empty() {
+            writeln!(
+                formatter,
+                "Removed {} edges from {} out of {} commits.",
+                stats.edges, stats.commits, orig_commits
+            )?;
+        }
+    }
+    tx.finish(ui, format!("simplified {} commits", orig_commits))?;
+
+    Ok(())
+}
+
+#[derive(Default)]
+struct SimplifyStats {
+    commits: usize,
+    edges: usize,
+}
+
+impl SimplifyStats {
+    fn for_commit(edges: usize) -> Self {
+        Self { commits: 1, edges }
+    }
+
+    fn add(&mut self, other: SimplifyStats) {
+        self.commits += other.commits;
+        self.edges += other.edges;
+    }
+
+    fn is_empty(&self) -> bool {
+        self.commits == 0 && self.edges == 0
+    }
+}
+
+fn simplify_commit(
+    settings: &UserSettings,
+    tx: &mut WorkspaceCommandTransaction,
+    commit: &Commit,
+) -> Result<SimplifyStats, CommandError> {
+    if commit.parent_ids().len() <= 1 {
+        return Ok(SimplifyStats::default());
+    }
+
+    let old_heads: HashSet<_> = commit.parent_ids().iter().cloned().collect();
+    let new_heads: HashSet<_> = tx
+        .repo_mut()
+        .index()
+        .heads(&mut commit.parent_ids().iter())
+        .into_iter()
+        .collect();
+    if new_heads == old_heads {
+        return Ok(SimplifyStats::default());
+    }
+    let removed = old_heads.len() - new_heads.len();
+
+    tx.repo_mut()
+        .rewrite_commit(settings, commit)
+        .set_parents(new_heads.into_iter().collect())
```

Comment:
I think you might reorder the untouched parents in an arbitrary order. Order should probably be preserved.